webapp/packages/core-blocks/src/useSearch.ts

```diff
@@ -0,0 +1,107 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { action, autorun, computed, observable } from 'mobx';
+
+import { isNotNullDefined } from '@cloudbeaver/core-utils';
+
+import { useFuzzySearch } from './useFuzzySearch.js';
+import { useObservableRef } from './useObservableRef.js';
+
+export type SearchStrategy = 'startsWith' | 'contains' | 'fuzzy';
+
+interface Props<T extends object> {
+  sourceHints: T[];
+  searchFields: (keyof T)[];
+  matchStrategy?: SearchStrategy;
+  predicate?: (suggestion: T, lastWord?: string) => boolean;
+}
+
+export interface UseSearchAPI<T> {
+  searchResult: T[];
+  setSearch: (searchWord: string) => void;
+}
+
+export function useSearch<T extends object>({
+  sourceHints,
+  searchFields,
+  predicate,
+  matchStrategy = 'contains',
+}: Props<T>): Readonly<UseSearchAPI<T>> {
+  const fuzzySearch = useFuzzySearch({
+    sourceProposals: sourceHints,
+    fields: searchFields as string[],
+  });
+  const state = useObservableRef(
+    () => ({
+      search: '',
+      setSearch(searchWord: string) {
+        if (this.matchStrategy === 'fuzzy') {
+          this.fuzzySearch.engine.search(searchWord, {
+            filter: result => {
+              const suggestion = this.sourceHints.find(suggestion =>
+                this.searchFields.every(field => result.terms.includes(suggestion[field] as string)),
+              );
+
+              if (!suggestion) {
+                return false;
+              }
+
+              return filterBase(suggestion, this.searchFields, searchWord, this.predicate);
+            },
+          });
+          return;
+        }
```

Comment:
Why do we use early return here? Is it expected that with a `fuzzy` strategy, `setSeach` will not change the `search` state?